dsros_sensors: depend on dave_gazebo_world_plugins and export library

mabelzhang opened the pull request:
Followup to the error in https://github.com/Field-Robotics-Lab/ds_sim/pull/5#issuecomment-1000125169
Depends on https://github.com/Field-Robotics-Lab/dave/pull/179

- Move `catkin_package()` up because according to http://wiki.ros.org/catkin/CMakeLists.txt , it must be called before any `add_library` and `add_executable`.
- Add `catkin_package(LIBRARIES dsros_sensors)`. This exports `dsros_sensors` library for downstream packages (`glider_hybrid_whoi_gazebo` for example) that depend on `ds_sim`. Make `dsros_sensors` `SHARED`.
- Add `${catkin_LIBRARIES}` dependency to `dsros_sensors` target. This gives `dsros_sensors` access to `dave_gazebo_world_plugins` which has been added to `catkin_package(CATKIN_DEPENDS)`. That is where symbol `StratifiedCurrentVelocity` is defined in a protobuf message.

Essentially, what these two PRs do is that they export `dave_gazebo_world_plugins` and `ds_sim` for downstream libraries, and links them together.

With these, I don't get that error anymore.

However, the manual `-s libdsros_sensors.so` flag in the `glider_hybrid_whoi_gazebo` launch files is still needed. It shouldn't be needed if things are set up correctly. I didn't try to track down where the launch file eventually leads to `dsros_sensors`. It isn't apparent to me from looking at the world files. Do we know where exactly the sensors are used in the glider worlds?
If we want to fix it, it would be in a separate PR. This PR is ready.

mabelzhang committed:
dsros_sensors: depend on dave_gazebo_world_plugins and export the lib

Signed-off-by: Mabel Zhang <mabel@openrobotics.org>

woensug-choi reviewed (approved):
With,  [this](https://github.com/Field-Robotics-Lab/ds_sim/pull/6) and https://github.com/Field-Robotics-Lab/dave/pull/179
Yes! It works now :D https://github.com/Field-Robotics-Lab/ds_sim/pull/5#issuecomment-1000125169 is solved!

woensug-choi commented:
> However, the manual -s libdsros_sensors.so flag in the glider_hybrid_whoi_gazebo launch files is still needed. It shouldn't be needed if things are set up correctly. I didn't try to track down where the launch file eventually leads to dsros_sensors. It isn't apparent to me from looking at the world files. Do we know where exactly the sensors are used in the glider worlds?
If we want to fix it, it would be in a separate PR. This PR is ready.

We are explicitly importing `dsros_sensors` manually with `-s libdsros_sensors.so` at dave repo's DVL demo launch too.
Look here,

https://github.com/Field-Robotics-Lab/dave/blob/cdc6b4d209c730f9637341bcabb6d7355adda84c/examples/dave_demo_launch/launch/dave_dvl_demo_dsl.launch#L20

mabelzhang committed:
fix include

Signed-off-by: Mabel Zhang <mabel@openrobotics.org>

mabelzhang commented:
I looked into the `-s` flag more, and now I'm not sure it's not needed. Maybe this is a thing with custom sensors that I wasn't aware of.

I tracked down one of the sensor worlds - `dave_dvl_gradient_demo_dsl.launch`, which loads `dave_sensor_models/models/teledyne_whn_dsl`, which loads the `dsros_dvl` plugin. I tried installing all the packages, exporting libraries in all the places, adding the `<export><gazebo_ros plugin_path="${prefix}/../../lib" /></export>` tags to `ds_sim`.

In the end, the flag was still needed. I also saw in a community package that's supposed to help make writing custom sensors easier, that they need the flag too https://github.com/peci1/gazebo_custom_sensor_preloader So maybe that's the thing to do for custom sensors.

The good thing that came out of this was that I did some cleanups in a few CMakeLists.txt / package.xml files in the dave repo, and running `install` on packages seems to work. I'll PR those separately.

As for `ds_sim`, do we know why the `LIBRARIES` line in `CMakeLists.txt` was commented out? In this PR, I only added back one library, but I'm inclined to just uncomment the whole line, because most likely anything depending on this package will want those libraries.

woensug-choi commented:
https://github.com/Field-Robotics-Lab/ds_sim/pull/6/commits/0604360f456a2974dd53ceb6300ed5f8e264c38b broke something. Rolling back to a commit before compiling ok.

 ```bash
Errors     << ds_sim:make /home/woensug/uuv_ws/logs/ds_sim/build.make.016.log
In file included from /home/woensug/uuv_ws/src/ds_sim/gazebo_src/dsros_dvl.cc:42:
/home/woensug/uuv_ws/src/ds_sim/gazebo_src/dsros_dvl.hh:60:10: fatal error: dave_gazebo_world_plugins/StratifiedCurrentVelocity.pb.h: No such file or directory
   60 | #include <dave_gazebo_world_plugins/StratifiedCurrentVelocity.pb.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/dsros_sensors.dir/build.make:102: CMakeFiles/dsros_sensors.dir/gazebo_src/dsros_dvl.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:3206: CMakeFiles/dsros_sensors.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
cd /home/woensug/uuv_ws/build/ds_sim; catkin build --get-env ds_sim | catkin env -si  /usr/bin/make --jobserver-auth=3,4; cd -
```

dtdavi1 commented:
The '-s' flag is required because the DVL is implemented as a sensor in ds_sim (as opposed to a plugin).

I'm not sure about the commented out LIBRARY line.  We haven't changed that, so it's an artifact of the original WHOI master branch.

mabelzhang commented:
> 0604360 broke something. Rolling back to a commit before compiling ok.

Fixed in https://github.com/Field-Robotics-Lab/dave/pull/182

So the reason I changed that `#include` line is that, it was inconsistent whether the user specifies `catkin config --install` and `catkin config --no-install`. The former wanted the package name in the `#include` line, the latter didn't. That's incorrect. The source files shouldn't be different between whether the package is installed or not.

The new PR above exports the `build` directory, as opposed to `build/dave_gazebo_world_plugins`. That way, when including the `.pb.h` file, the `#include` line has to specify the package name. That is the recommended way because, if you have multiple messages of the same name in different packages, this makes sure they don't conflict.

mabelzhang committed:
fix typo

Signed-off-by: Mabel Zhang <mabel@openrobotics.org>

mabelzhang committed:
uncomment LIBRARIES line

Signed-off-by: Mabel Zhang <mabel@openrobotics.org>

mabelzhang commented:
> The '-s' flag is required because the DVL is implemented as a sensor in ds_sim (as opposed to a plugin).

I asked around internally to see if we have a best practice alternative for that, and turns out our users are better experts than us at breaking through Gazebo's limits that make custom sensors difficult. It seems that `-s` is the way to do it, though it is what one might call a hack around what Gazebo isn't designed to do.

> I'm not sure about the commented out LIBRARY line. We haven't changed that, so it's an artifact of the original WHOI master branch.

Okay I'll go ahead and uncomment that line. b67a07c

woensug-choi commented:
Happy to report that my problems are solved!
